Common/Data/FileSync.swift

```diff
@@ -0,0 +1,253 @@
+    }
+
+    fileprivate func downloadContent(of directory: File) -> Future<[String: Any], SCError> {
+        guard directory.isDirectory else { return Future(error: .other("only works on directory") ) }
+
+        let request = self.request(for: getQueryURL(for: directory)! )
+        let promise: Promise<[String: Any], SCError> = Promise()
+        metadataSession.dataTask(with: request) { data, response, error in
+            var responseData: Data
+            do {
+                responseData = try self.confirmNetworkResponse(data: data, response: response, error: error)
+            } catch let error as SCError {
+                promise.failure(error)
+                return
+            } catch {
+                promise.failure( .other("Weird"))
+                return
+            }
+
+            guard let json = (try? JSONSerialization.jsonObject(with: responseData, options: .allowFragments)) as? [String: Any] else {
+                promise.failure(SCError.jsonDeserialization("Can't deserialize"))
+                return
+            }
+
+            promise.success(json)
+            }.resume()
+        return promise.future
+    }
+
+    public func downloadSharedFiles() -> Future<[[String: Any]], SCError> {
+        let promise = Promise<[[String: Any]], SCError>()
+
+        let request = self.request(for: Brand.default.servers.backend.appendingPathComponent("files") )
+        metadataSession.dataTask(with: request) { data, response, error in
+            do {
+                let data = try self.confirmNetworkResponse(data: data, response: response, error: error)
+                let json = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as! MarshaledObject
+                let files: [MarshaledObject] = try json.value(for: "data")
+                let sharedFiles = files.filter { object -> Bool in
+                    return (try? object.value(for: "context")) == "geteilte Datei"
+                }
+
+                promise.success(sharedFiles as! [[String: Any]])
+            } catch let error as SCError {
+                promise.failure(error)
+            } catch let error {
+                promise.failure(.jsonDeserialization(error.localizedDescription) )
+            }
+            }.resume()
+        return promise.future
+    }
+
+    // MARK: File materialization
+    public func download(_ file: File,
+                         background: Bool = false,
+                         onDownloadInitialised: @escaping () -> Void,
+                         progressHandler: @escaping ProgressHandler ) -> Future<URL, SCError> {
+        let localURL = file.localURL
+        let downloadSession: URLSession = background ? self.backgroundSession : self.foregroundSession
+        guard [File.DownloadState.notDownloaded, File.DownloadState.downloadFailed].contains(file.downloadState) else {
+            return Future<URL, SCError>(value: localURL)
+        }
+
+        file.downloadState = .downloading
```

Comment:
It might be possible that a download is initiated for a download that is currently downloading. e.g. by the file provider

Reply:
After some thought, this should be checked on the callers site. There is no unified way of dealing with this situation that are good that i can think of.